### ISSUE TITLE
cmd/go: go build internal compiler error

### ISSUE THREAD

**Issue by toy80** [golang/go#53559]
### What version of Go are you using (`go version`)?

<pre>
go version go1.18.2 linux/amd64
go version go1.18 windows/amd64
</pre>

### Does this issue reproduce with the latest release?
yes

### What operating system and processor architecture are you using (`go env`)?

<details><summary><code>go env</code> Output</summary><br><pre>
$ go env
GO111MODULE="on"
GOARCH="amd64"
GOBIN=""
GOEXE=""
GOEXPERIMENT=""
GOFLAGS=""
GOHOSTARCH="amd64"
GOHOSTOS="linux"
GOINSECURE=""
GONOPROXY=""
GONOSUMDB="*"
GOOS="linux"
GOPRIVATE=""
GOPROXY=
GOROOT="/usr/local/go"
GOSUMDB="sum.golang.org"
GOTMPDIR=""
GOTOOLDIR="/usr/local/go/pkg/tool/linux_amd64"
GOVCS=""
GOVERSION="go1.18.2"
GCCGO="gccgo"
GOAMD64="v1"
AR="ar"
CC="gcc"
CXX="g++"
CGO_ENABLED="1"
GOMOD="/c/w/compiler_error/go.mod"
GOWORK=""
CGO_CFLAGS="-g -O2"
CGO_CPPFLAGS=""
CGO_CXXFLAGS="-g -O2"
CGO_FFLAGS="-g -O2"
CGO_LDFLAGS="-g -O2"
PKG_CONFIG="pkg-config"
GOGCCFLAGS="-fPIC -m64 -pthread -fmessage-length=0 -fdebug-prefix-map=/tmp/go-build2088256825=/tmp/go-build -gno-record-gcc-switches"

</pre></details>

### What did you do?

I made a test case, build it with -race and -gcflags=all="-l -N", then an "internal compiler error" occurs.

```go
package bar

type A[T any] struct {
	X, Y int
}

func (A[T]) Wtf() {}

func Bar[T any]() any {
	return A[T]{}
}

```

```go
package foo

import (
	"test1/bar"
)

func Foo() {
	bar.Bar[int]()
}

```

```
module test1

go 1.18

```

```
go build -race -gcflags=all="-l -N" ./foo
```

### What did you expect to see?
nothing

### What did you see instead?

```
test1/foo
<autogenerated>:1: internal compiler error: dictionary for A[go.shape.int_0].Wtf should only use concrete types: go.shape.int_0

Please file a bug report including a short program that triggers the error.
https://go.dev/issue/new
```

**Comment by ZekeLu** [comment]
This issue has been fixed by 5b4fafd5db1c13e5e2c9f836476015870b3ed30b

**Comment by ianlancetaylor** [comment]
Thanks.
